FarringtonFlexible with weekly data results in RRuntimeError

I've found that when using daily case count data, I can use `FarringtonFlexible` without problem but with weekly or multi-week data, I am getting a runtime error.  I've tried all sorts of ways of working around this including passing in `glmWarnings` as `False` to try to avoid since this it looks more like a `warning()` than an actual failure.

I will keep trying to debug this on my own, but I'm not a great R programmer and it's hard to tell if this is an issue in the `surveillance` package or how I am constructing data or something else.

So while I will keep trying to troubleshoot this, a few questions if there are any thoughts:
1. Would it help if I could try to reproduce this in R?
2. Can you think of any other workarounds?
3. Specifically, do you know if there is any way in `rpy2` or any other way of ignoring an issue like this since it looks more like a `warning()` than a fatal `Exception`?

RRuntimeError (a fuller callstack is below):

```
E       rpy2.rinterface.RRuntimeError: Error in algo.farrington.data.glm(dayToConsider = dayToConsider, b = control$b,  : 
E         Some reference values did not exist (index<1).
```

Code to reproduce (as a `pytest`)

```
import pandas as pd
import pytest

from epysurv.models.timepoint import FarringtonFlexible

def test_farrington_weekly_example():

    model = FarringtonFlexible()

    total_periods = 100
    test_size = 20
    case_count = 10

    # set up some weekly data
    dates = pd.date_range('2017-07-09', periods=total_periods, freq='7D')
    # just make this constant (but this also fails with random or real case count values)
    case_counts = [case_count] * total_periods

    df = pd.DataFrame({'n_cases': case_counts}, index = dates)

    train_data = df[:-1 * test_size]
    test_data = df[-1 * test_size:]

    # make sure we can fit and predict
    model.fit(train_data)
    _ = model.predict(test_data)
```

Fuller callstack:

```
def test_farrington_weekly_example():
    
        model = FarringtonFlexible()
    
        total_periods = 100
        test_size = 20
        case_count = 10
    
        # set up some weekly data
        dates = pd.date_range('2017-07-09', periods=total_periods, freq='7D')
        # just make this constant (but this also fails with random or real case count values)
        case_counts = [case_count] * total_periods
    
        df = pd.DataFrame({'n_cases': case_counts}, index = dates)
    
        train_data = df[:-1 * test_size]
        test_data = df[-1 * test_size:]
    
        # make sure we can fit and predict
        model.fit(train_data)
>       _ = model.predict(test_data)

test_farrington_specific_example.py:26: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
..\epysurv\models\timepoint\_base.py:131: in predict
    surveillance_result = self._call_surveillance_algo(r_instance, detection_range)
..\epysurv\models\timepoint\farrington.py:166: in _call_surveillance_algo
    surv = surveillance.farringtonFlexible(sts, control=control)
C:\anaconda3\envs\epysurv-dev\lib\site-packages\rpy2\robjects\functions.py:178: in __call__
    return super(SignatureTranslatedFunction, self).__call__(*args, **kwargs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = R object with classes: ('function',) mapped to:
<DocumentedSTFunction - Python:0x000001C714FD1608 / R:0x000001C70F26E4A8>
args = (R object with classes: ('sts',) mapped to:
<RS4 - Python:0x000001C7151F5548 / R:0x000001C70FF061F0>,)
kwargs = {'control': R object with classes: ('list',) mapped to:
<ListVector - Python:0x000001C7152E92C8 / R:0x000001C71400440...ect with classes: ('character',) mapped to:
<StrVector - Python:0x000001C715468C88 / R:0x000001C7149A7E00>
['delta']}
new_args = [R object with classes: ('sts',) mapped to:
<RS4 - Python:0x000001C7151F5548 / R:0x000001C70FF061F0>]
new_kwargs = {'control': R object with classes: ('list',) mapped to:
<ListVector - Python:0x000001C7152E92C8 / R:0x000001C71400440...ect with classes: ('character',) mapped to:
<StrVector - Python:0x000001C715471C88 / R:0x000001C7149A7E00>
['delta']}
k = 'control'
v = R object with classes: ('list',) mapped to:
<ListVector - Python:0x000001C7152E92C8 / R:0x000001C714004408>
[IntVect...ject with classes: ('character',) mapped to:
<StrVector - Python:0x000001C715478F48 / R:0x000001C7149A7E00>
['delta']

    def __call__(self, *args, **kwargs):
        new_args = [conversion.py2ri(a) for a in args]
        new_kwargs = {}
        for k, v in kwargs.items():
            new_kwargs[k] = conversion.py2ri(v)
>       res = super(Function, self).__call__(*new_args, **new_kwargs)
E       rpy2.rinterface.RRuntimeError: Error in algo.farrington.data.glm(dayToConsider = dayToConsider, b = control$b,  : 
E         Some reference values did not exist (index<1).
```

--- a/.pre-commit-config.yaml
+++ b/.pre-commit-config.yaml
@@ -4,7 +4,6 @@ repos:
     rev: "v5.9.3"
     hooks:
       - id: isort
-        additional_dependencies: ["isort[pyproject]"]
   - repo: https://github.com/psf/black
     rev: 21.9b0
     hooks:
@@ -17,5 +16,5 @@ repos:
     rev: 3.9.2
     hooks:
       - id: flake8
-        exclude: ^(tests|examples|docs)/*
+        exclude: ^(tests|examples|docs)/.*
         additional_dependencies: [pep8-naming, flake8-docstrings]

--- a/.pre-commit-config.yaml
+++ b/.pre-commit-config.yaml
@@ -1,12 +1,12 @@
 fail_fast: true
 repos:
   - repo: https://github.com/pre-commit/mirrors-isort
-    rev: "v4.3.21"
+    rev: "v5.9.3"
     hooks:
       - id: isort
         additional_dependencies: ["isort[pyproject]"]
   - repo: https://github.com/psf/black
-    rev: stable
+    rev: 21.9b0
     hooks:
       - id: black
   - repo: https://github.com/pre-commit/mirrors-mypy
@@ -14,7 +14,7 @@ repos:
     hooks:
       - id: mypy
   - repo: https://gitlab.com/pycqa/flake8
-    rev: 3.7.8
+    rev: 3.9.2
     hooks:
       - id: flake8
         exclude: ^(tests|examples|docs)/*

--- a/epysurv/models/timepoint/_base.py
+++ b/epysurv/models/timepoint/_base.py
@@ -178,13 +178,12 @@ class STSBasedAlgorithm(SurveillanceRPackageAlgorithm):
 
         sts = surveillance.sts(
             start=r.c(data.index[0].year, _get_start_epoch(data)),
-            epoch=robjects.IntVector(
+            epoch=robjects.DateVector(
                 [
                     r["as.numeric"](r["as.Date"](d.isoformat()))[0]
                     for d in data.index.date
                 ]
             ),
-            # epoch=data.index,
             freq=_get_freq(data),
             observed=data["n_cases"].values,
             epochAsDate=True,

--- a/epysurv/models/timepoint/farrington.py
+++ b/epysurv/models/timepoint/farrington.py
@@ -1,5 +1,7 @@
 from dataclasses import dataclass
 
+import numpy as np
+import pandas as pd
 from rpy2.robjects import r
 from rpy2.robjects.packages import importr
 
@@ -148,6 +150,8 @@ class FarringtonFlexible(STSBasedAlgorithm):
     threshold_method: str = "delta"
 
     def _call_surveillance_algo(self, sts, detection_range):
+
+        self._check_enough_reference_data_available(detection_range, sts)
         control = r.list(
             range=detection_range,
             b=self.years_back,
@@ -165,3 +169,19 @@ class FarringtonFlexible(STSBasedAlgorithm):
 
         surv = surveillance.farringtonFlexible(sts, control=control)
         return surv
+
+    def _check_enough_reference_data_available(self, detection_range, sts):
+        dates = pd.to_datetime(np.asarray(sts.slots["epoch"]), unit="D")
+        py_detection_range = np.asarray(detection_range) - 1
+        prediction_dates = dates[py_detection_range]
+        required_reference_data = (
+            prediction_dates - (np.timedelta64(1, "Y") * self.years_back)
+        ).floor("D")
+        available_reference_data = dates[: (py_detection_range)[0]]
+        if not available_reference_data.min() < required_reference_data.min():
+            raise ValueError(
+                f"You are trying to use reference data from {self.years_back} years back for predictions "
+                f"starting from {prediction_dates[0].date()}, but the "
+                f"reference data goes only back until {available_reference_data.min().date()}. "
+                "Try setting a lower value for `years_back` or provide more reference data."
+            )

--- a/requirements-dev.txt
+++ b/requirements-dev.txt
@@ -4,8 +4,9 @@ pytest
 pytest-datadir
 pytest-cov
 codecov
-black==19.10b0
+black==21.9b0
 mypy==0.910
+isort==5.9.3
 flake8
 pep8-naming
 flake8-docstrings

--- a/tests/conftest.py
+++ b/tests/conftest.py
@@ -26,7 +26,10 @@ def test_data(shared_datadir):
 
 def _load_data(filepath):
     data = pd.read_csv(
-        filepath, index_col=0, parse_dates=True, infer_datetime_format=True,
+        filepath,
+        index_col=0,
+        parse_dates=True,
+        infer_datetime_format=True,
     )
     data.index.freq = pd.infer_freq(data.index)
     return data

--- a/tests/test_timepoint_models.py
+++ b/tests/test_timepoint_models.py
@@ -120,3 +120,27 @@ def test_prediction_witout_labels(train_data, test_data, shared_datadir, Algo):
     saved_predictions = drop_column_if_exists(saved_predictions, "upperbound")
 
     assert_frame_equal(pred, saved_predictions)
+
+
+def test_farrington_flexible__raises_on_too_less_reference_data():
+
+    model = FarringtonFlexible()
+
+    total_periods = 100
+    test_size = 20
+    case_count = 10
+
+    dates = pd.date_range("2017-07-09", periods=total_periods, freq="W")
+    case_counts = [case_count] * total_periods
+
+    df = pd.DataFrame({"n_cases": case_counts}, index=dates)
+
+    train_data = df[: -1 * test_size]
+    test_data = df[-1 * test_size :]
+
+    model.fit(train_data)
+    with pytest.raises(
+        ValueError,
+        match="You are trying to use reference data from 3 years back for predictions starting from 2019-01-20",
+    ):
+        _ = model.predict(test_data)

